WIP: Added anonymized location extension

kbgg opened the pull request:
**Related Issue(s):** #843


**Proposed Changes:**

1. Adding an "anonymized location" extension

**PR Checklist:**

- [x] This PR is made against the dev branch (all proposed changes except releases should be against dev, not master).
- [x] This PR has **no** breaking changes.
- [ ] I have added my changes to the [CHANGELOG](https://github.com/radiantearth/stac-spec/blob/dev/CHANGELOG.md) **or** a CHANGELOG entry is not required.
- [ ] This PR affects the [STAC API spec](https://github.com/radiantearth/stac-api-spec), and I have opened issue/PR #XXX to track the change.

At Radiant we've come across instances where we can not provide or do not have precise locations of labeled data. A solution we've come to is defining a grid across the globe and assigning each item to the grid square which it falls within. The geometry of the STAC item will then be set to that grid square instead of its actual location. Most of the time we just create 1x1 degree grid squares and that provides enough anonymization and geographic accuracy to still be able to train models.

There should still be some way for the user to know that the geometry present in the STAC item is not the actual geometry of the label item but is instead a grid square of where the item *could* be. The anonymized location extension being present would be an indication that the geometry should not be taken as true and the anon:precision property would indicate how large the grid squares are in terms of decimal degrees.

This could also solve the problem of data which has no geometry (#843). Instead of setting the geometry property to null you can define the grid as having only one grid square which covers the entire globe and the item falls within that one Earth-sized grid square.

This is still a WIP so any feedback on this solution is welcomed.

cholmes commented on extensions/anonymized-location/README.md:
```diff
@@ -0,0 +1,33 @@
+
+| Field Name       | Type                 | Description |
+| ---------------- | -------------------- | ----------- |
+| anon:precision   | number               | The maximum precision of the latitude and longitude coordinates present in the bbox and geometry properties. |
```
Should probably mark this as required. Also I think it'd be good to have a bit more detail of what the number actually means. 1 means 1 degree square grid? 2 means 2 degrees? How does that relate to the number of significant digits? Like if my box is 148.25, 59.25, 152.25, 60.25 is that '1' because it's a 1x1 sized area? I think any answer is fine here, just should have more details to help people figure out the value they should put here.

Also this field should probably be marked as 'required'.

cholmes commented on extensions/anonymized-location/README.md:
```diff
@@ -0,0 +1,33 @@
+| ---------------- | -------------------- | ----------- |
+| anon:precision   | number               | The maximum precision of the latitude and longitude coordinates present in the bbox and geometry properties. |
+
+*At least one of the fields must be specified.*
```
This line is a bit odd if there's only one field.

cholmes commented on extensions/anonymized-location/examples/collection.json:
```diff
@@ -0,0 +1,34 @@
+{
+  "stac_version": "1.0.0-beta.1",
```
should update to beta2

m-mohr replied:
Also, the JSON Schema should be updated to reflect recent changes we did to extension schemas.

cholmes commented on extensions/anonymized-location/examples/item.json:
```diff
@@ -0,0 +1,241 @@
+            "type": "text/plain"
+        },
+        "B1": {
+            "href": "https://landsat-pds.s3.amazonaws.com/c1/L8/107/018/LC08_L1TP_107018_20181001_20181001_01_RT/LC08_L1TP_107018_20181001_20181001_01_RT_B1.TIF",
```
This example confuses me a bit, since landsat data from the public bucket pretty easily lets you get its location. I think it'd be better to use a more 'real' example of some image that is commercial / private.

cholmes replied:
Like the assets linked should probably not be a geotiff, since that would have the actual location in it.

kbgg commented:
> I put in a few individual comments. Overall I think the approach makes sense. Though I do worry that people may not notice that the data they're using is anonymized and may get confused. They'll need to know to look for 'anon:precision', and if they've not come across this extension before it's not obvious what that will be.
> 
> I find myself wanting some warning directly in the item / collection. Like a requirement in the description that says something like 'the geometry is not accurate since data has been anonymized.' Or a required 'anon:warning' field that has a value "this data has been anonymized - the geometry is only an approximation of the exact location'. Another route could be to only use the bbox and not the geometry, now that we can define null geometries. Or I'm open to other possibilities, but I think we have to give some clearer sign to people who may just be searching for general STAC data and be confused by it being not the actual footprint.

Thank you for all your feedback!

I've update the anon:precision field to anon:size and added a new description to better reflect what the field is supposed to contain. I've also added an anon:warning field which should let users know that the location is only approximate. Examples are also updated to be a sample of the dataset we intend to use this extension with.

matthewhanson commented:
As extensions now live in the stac-extensions organization, this PR is being closed. Follow instructions here to create a repository for the extension:
https://github.com/stac-extensions/stac-extensions.github.io#adding-a-new-extension
